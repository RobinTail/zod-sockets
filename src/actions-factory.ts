import { z } from "zod";
import { Action, Handler } from "./action";
import { EmissionMap } from "./emission";
import { SocketsConfig } from "./sockets";

export interface SimpleActionDef<
  IN extends z.AnyZodTuple,
  E extends EmissionMap,
> {
  input: IN;
  handler: Handler<z.output<IN>, void, E>;
}

export interface AckActionDef<
  IN extends z.AnyZodTuple,
  OUT extends z.AnyZodTuple,
  E extends EmissionMap,
> {
  input: IN;
  output: OUT;
  handler: Handler<z.output<IN>, z.input<OUT>, E>;
}

export class ActionsFactory<E extends EmissionMap> {
  constructor(protected config: SocketsConfig<E>) {}

  public build<IN extends z.AnyZodTuple, OUT extends z.AnyZodTuple>(
    def: SimpleActionDef<IN, E> | AckActionDef<IN, OUT, E>,
  ): Action<IN, OUT> {
    return new Action(def);
  }
}
