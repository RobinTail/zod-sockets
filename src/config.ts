import { EmissionMap } from "./emission";
import { AbstractLogger } from "./logger";
import {
  FallbackNamespaces,
  Namespace,
  Namespaces,
  RootNS,
  fallbackNamespaces,
  rootNS,
} from "./namespace";

interface ConstructorOptions<NS extends Namespaces> {
  /**
   * @desc The instance of a logger
   * @default console
   * */
  logger?: AbstractLogger;
  /**
   * @desc The acknowledgment awaiting timeout
   * @default 2000
   * */
  timeout?: number;
  /**
   * @desc You can disable the startup logo.
   * @default true
   */
  startupLogo?: boolean;
  /**
   * @desc Define namespaces inline or consider using addNamespace() method
   * @default {}
   * @see Namespace
   * */
  namespaces?: NS;
}

export class Config<T extends Namespaces> {
  public readonly logger: AbstractLogger;
  public readonly timeout: number;
  public readonly startupLogo: boolean;
  public readonly namespaces: T;

  public constructor({
    logger = console,
    timeout = 2000,
    startupLogo = true,
    namespaces = fallbackNamespaces as unknown as T,
  }: ConstructorOptions<T>) {
    this.logger = logger;
    this.timeout = timeout;
    this.startupLogo = startupLogo;
    this.namespaces = namespaces;
  }

  public addNamespace<E extends EmissionMap = {}, K extends string = RootNS>({
    path = rootNS as K,
    emission = {} as E,
    hooks = {},
  }: Partial<Namespace<E>> & { path?: K }): Config<
    Omit<T, K> & Record<K, Namespace<E>>
  > {
    const { logger, timeout, startupLogo, namespaces } = this;
    return new Config({
      logger,
      timeout,
      startupLogo,
      namespaces: { ...namespaces, [path]: { emission, hooks } },
    });
  }
}

export const createConfig = <T extends Namespaces = FallbackNamespaces>(
  def: ConstructorOptions<T> = {},
) => new Config(def);
