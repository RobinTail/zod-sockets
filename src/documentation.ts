import { ContactObject, LicenseObject } from "openapi3-ts/oas31";
import { z } from "zod";
import { AbstractAction } from "./action";
import { AsyncApiDocumentBuilder } from "./async-api/document-builder";
import { AsyncChannelObject } from "./async-api/commons";
import { Config } from "./config";
import { depicters, onEach, onMissing } from "./documentation-helpers";
import { Namespaces, normalizeNS } from "./namespace";
import { walkSchema } from "./schema-walker";

interface DocumentationParams {
  title: string;
  version: string;
  documentId?: string;
  description?: string;
  contact?: ContactObject;
  license?: LicenseObject;
  servers?: Record<string, { url: string }>;
  actions: AbstractAction[];
  config: Config<Namespaces>;
}

export class Documentation extends AsyncApiDocumentBuilder {
  public constructor({
    actions,
    config: { namespaces },
    title,
    version,
    documentId,
    description,
    contact,
    license,
    servers = {},
  }: DocumentationParams) {
    super({
      info: { title, version, contact, license, description },
      id: documentId,
      defaultContentType: "text/plain",
    });
    for (const server in servers) {
      this.addServer(server, { ...servers[server], protocol: "socket.io" });
      if (!this.document.id) {
        const uri = new URL(servers[server].url.toLowerCase());
        this.document.id = `urn:${uri.host.split(".").concat(uri.pathname.slice(1).split("/")).join(":")}`;
      }
    }
    for (const [ns, { emission }] of Object.entries(namespaces)) {
      const channel: AsyncChannelObject = {
        description: `${normalizeNS(ns)} namespace`,
        bindings: {
          "socket.io": {
            bindingVersion: "0.11.0",
            method: "GET",
            headers: walkSchema({
              direction: "in",
              schema: z.object({
                connection: z.literal("Upgrade").optional(),
                upgrade: z.literal("websocket").optional(),
              }),
              onEach,
              onMissing,
              rules: depicters,
            }),
            query: {
              ...walkSchema({
                direction: "in",
                schema: z.object({
                  EIO: z
                    .literal("4")
                    .describe("Mandatory, the version of the protocol"),
                  transport: z
                    .union([z.literal("polling"), z.literal("websocket")])
                    .describe("Mandatory, the name of the transport."),
                  sid: z.string().optional().describe("The session identifier"),
                }),
                onEach,
                onMissing,
                rules: depicters,
              }),
              externalDocs: {
                description: "Engine.IO Protocol",
                url: "https://socket.io/docs/v4/engine-io-protocol/",
              },
            },
          },
        },
        subscribe: {
          operationId: `out${normalizeNS(ns)}`,
          description: `The messages produced by the application within the ${normalizeNS(ns)} namespace`,
          message: {
            oneOf: Object.entries(emission).map(([event, { schema, ack }]) => ({
              name: event,
              title: event,
              messageId: `out${normalizeNS(ns)}/${event}`,
              payload: walkSchema({
                direction: "out",
                schema,
                onEach,
                onMissing,
                rules: depicters,
              }),
              bindings: ack
                ? {
                    "socket.io": {
                      bindingVersion: "0.11.0",
                      ack: walkSchema({
                        direction: "in",
                        schema: ack.describe(
                          ack.description || "Acknowledgement",
                        ),
                        onEach,
                        onMissing,
                        rules: depicters,
                      }),
                    },
                  }
                : undefined,
            })),
          },
        },
        publish: {
          operationId: `in${normalizeNS(ns)}`,
          description: `The messages consumed by the application within the ${normalizeNS(ns)} namespace`,
          message: {
            oneOf: actions
              .filter((action) => action.getNamespace() === ns)
              .map((action) => {
                const event = action.getEvent();
                const output = action.getSchema("output");
                return {
                  name: event,
                  title: event,
                  messageId: `in${normalizeNS(ns)}/${event}`,
                  payload: walkSchema({
                    direction: "in",
                    schema: action.getSchema("input"),
                    onEach,
                    onMissing,
                    rules: depicters,
                  }),
                  bindings: output
                    ? {
                        "socket.io": {
                          bindingVersion: "0.11.0",
                          ack: walkSchema({
                            direction: "out",
                            schema: output.describe(
                              output.description || "Acknowledgement",
                            ),
                            onEach,
                            onMissing,
                            rules: depicters,
                          }),
                        },
                      }
                    : undefined,
                };
              }),
          },
        },
      };
      this.addChannel(ns, channel);
    }
  }
}
