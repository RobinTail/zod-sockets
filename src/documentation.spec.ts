import { config as exampleConfig } from "../example/config";
import { actions } from "../example/actions";
import { ActionsFactory } from "./actions-factory";
import { createConfig } from "./config";
import { Documentation } from "./documentation";
import { z } from "zod";
import { describe, expect, test } from "vitest";

describe("Documentation", () => {
  const sampleConfig = createConfig();
  const factory = new ActionsFactory(sampleConfig);

  describe("Basic cases", () => {
    test("should generate the correct schema of the example", () => {
      const spec = new Documentation({
        actions,
        config: exampleConfig,
        version: "1.2.3",
        title: "Example API",
        servers: {
          example: { url: "https://example.com/socket.io" },
        },
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test("should generate the correct schema for complex types", () => {
      const literalValue = "something" as const;
      const spec = new Documentation({
        config: sampleConfig,
        actions: [
          factory.build({
            event: "test",
            input: z.tuple([
              z.array(z.number().int().positive()).min(1).max(3),
              z.array(z.boolean()),
              z.string().transform((str) => str.length),
            ]),
            output: z.tuple([z.literal(literalValue), z.number()]),
            handler: async ({ input }) => [literalValue, input[2]],
          }),
        ],
        version: "3.4.5",
        title: "Testing Complex Types",
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test("should generate the correct schema for nullable and optional types", () => {
      const spec = new Documentation({
        config: sampleConfig,
        actions: [
          factory.build({
            event: "test",
            input: z.tuple([
              z.string().optional(),
              z.string().optional().default("test"),
              z.boolean().nullish(),
              z.number().int().positive().nullish().default(123),
            ]),
            output: z.tuple([z.string().nullable()]),
            handler: async () => [null],
          }),
        ],
        version: "3.4.5",
        title: "Testing Nullable and Optional Types",
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    /*
    test("should generate the correct schema for intersection type", () => {
      const spec = new Documentation({
        config: sampleConfig,
        actions: [
          factory.build({
            event: "test",
            input: z.tuple([
              z.intersection(
                z.object({
                  one: z.string(),
                }),
                z.object({
                  two: z.string(),
                }),
              ),
            ]),
            output: z.tuple([
              z
                .object({
                  five: z.number().int().gte(0),
                })
                .and(
                  z.object({
                    six: z.string(),
                  }),
                ),
            ]),
            handler: async () => ({ five: 5, six: "six" }),
          }),
        ],
        version: "3.4.5",
        title: "Testing Intersection and And types",
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });*/
  });
});
