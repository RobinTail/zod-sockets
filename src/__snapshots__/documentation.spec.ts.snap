// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Documentation > Basic cases > should generate the correct schema for complex types 1`] = `
"asyncapi: 2.5.0
info:
  title: Testing Complex Types
  version: 3.4.5
defaultContentType: text/plain
tags: []
servers: {}
channels:
  /:
    description: / namespace
    bindings:
      socket.io:
        bindingVersion: 0.11.0
        method: GET
        headers:
          type: object
          properties:
            connection:
              type: string
              enum:
                - Upgrade
            upgrade:
              type: string
              enum:
                - websocket
        query:
          type: object
          properties:
            EIO:
              type: string
              enum:
                - "4"
              description: Mandatory, the version of the protocol
            transport:
              oneOf:
                - type: string
                  enum:
                    - polling
                - type: string
                  enum:
                    - websocket
              description: Mandatory, the name of the transport.
            sid:
              type: string
              description: The session identifier
          required:
            - EIO
            - transport
          externalDocs:
            description: Engine.IO Protocol
            url: https://socket.io/docs/v4/engine-io-protocol/
    subscribe:
      operationId: out/
      description: The messages produced by the application within the / namespace
      message:
        oneOf: []
    publish:
      operationId: in/
      description: The messages consumed by the application within the / namespace
      message:
        oneOf:
          - name: test
            title: test
            messageId: in//test
            payload:
              type: object
              format: tuple
              additionalProperties: false
              properties:
                "0":
                  type: array
                  items:
                    type: integer
                    format: int64
                    exclusiveMinimum: 0
                    maximum: 9007199254740991
                  minItems: 1
                  maxItems: 3
                "1":
                  type: array
                  items:
                    type: boolean
                "2":
                  type: string
            bindings:
              socket.io:
                bindingVersion: 0.11.0
                ack:
                  type: object
                  format: tuple
                  additionalProperties: false
                  properties:
                    "0":
                      type: string
                      enum:
                        - something
                    "1":
                      type: number
                      format: double
                      minimum: -1.7976931348623157e+308
                      maximum: 1.7976931348623157e+308
components: {}
"
`;

exports[`Documentation > Basic cases > should generate the correct schema for nullable and optional types 1`] = `
"asyncapi: 2.5.0
info:
  title: Testing Nullable and Optional Types
  version: 3.4.5
defaultContentType: text/plain
tags: []
servers: {}
channels:
  /:
    description: / namespace
    bindings:
      socket.io:
        bindingVersion: 0.11.0
        method: GET
        headers:
          type: object
          properties:
            connection:
              type: string
              enum:
                - Upgrade
            upgrade:
              type: string
              enum:
                - websocket
        query:
          type: object
          properties:
            EIO:
              type: string
              enum:
                - "4"
              description: Mandatory, the version of the protocol
            transport:
              oneOf:
                - type: string
                  enum:
                    - polling
                - type: string
                  enum:
                    - websocket
              description: Mandatory, the name of the transport.
            sid:
              type: string
              description: The session identifier
          required:
            - EIO
            - transport
          externalDocs:
            description: Engine.IO Protocol
            url: https://socket.io/docs/v4/engine-io-protocol/
    subscribe:
      operationId: out/
      description: The messages produced by the application within the / namespace
      message:
        oneOf: []
    publish:
      operationId: in/
      description: The messages consumed by the application within the / namespace
      message:
        oneOf:
          - name: test
            title: test
            messageId: in//test
            payload:
              type: object
              format: tuple
              additionalProperties: false
              properties:
                "0":
                  type: string
                "1":
                  type: string
                  default: test
                "2":
                  type:
                    - boolean
                    - "null"
                "3":
                  type:
                    - integer
                    - "null"
                  format: int64
                  exclusiveMinimum: 0
                  maximum: 9007199254740991
                  default: 123
            bindings:
              socket.io:
                bindingVersion: 0.11.0
                ack:
                  type: object
                  format: tuple
                  additionalProperties: false
                  properties:
                    "0":
                      type:
                        - string
                        - "null"
components: {}
"
`;

exports[`Documentation > Basic cases > should generate the correct schema of the example 1`] = `
"asyncapi: 2.5.0
info:
  title: Example API
  version: 1.2.3
id: urn:example:com:socket.io
defaultContentType: text/plain
tags: []
servers:
  example:
    url: https://example.com/socket.io
    protocol: socket.io
channels:
  /:
    description: / namespace
    bindings:
      socket.io:
        bindingVersion: 0.11.0
        method: GET
        headers:
          type: object
          properties:
            connection:
              type: string
              enum:
                - Upgrade
            upgrade:
              type: string
              enum:
                - websocket
        query:
          type: object
          properties:
            EIO:
              type: string
              enum:
                - "4"
              description: Mandatory, the version of the protocol
            transport:
              oneOf:
                - type: string
                  enum:
                    - polling
                - type: string
                  enum:
                    - websocket
              description: Mandatory, the name of the transport.
            sid:
              type: string
              description: The session identifier
          required:
            - EIO
            - transport
          externalDocs:
            description: Engine.IO Protocol
            url: https://socket.io/docs/v4/engine-io-protocol/
    subscribe:
      operationId: out/
      description: The messages produced by the application within the / namespace
      message:
        oneOf:
          - name: time
            title: time
            messageId: out//time
            payload:
              type: object
              format: tuple
              additionalProperties: false
              properties:
                "0":
                  type: string
                  description: current ISO time
          - name: chat
            title: chat
            messageId: out//chat
            payload:
              type: object
              format: tuple
              additionalProperties: false
              properties:
                "0":
                  type: string
                  description: message
                "1":
                  type: object
                  properties:
                    from:
                      type: string
                      description: the ID of author
                  required:
                    - from
                  description: extra info
          - name: rooms
            title: rooms
            messageId: out//rooms
            payload:
              type: object
              format: tuple
              additionalProperties: false
              properties:
                "0":
                  type: array
                  items:
                    type: string
                  description: room IDs
    publish:
      operationId: in/
      description: The messages consumed by the application within the / namespace
      message:
        oneOf:
          - name: chat
            title: chat
            messageId: in//chat
            payload:
              type: object
              format: tuple
              additionalProperties: false
              properties:
                "0":
                  type: string
                  description: message
          - name: ping
            title: ping
            messageId: in//ping
            payload:
              type: object
              format: tuple
              additionalProperties:
                format: any
                description: Anything
              properties: {}
            bindings:
              socket.io:
                bindingVersion: 0.11.0
                ack:
                  type: object
                  format: tuple
                  additionalProperties:
                    format: any
                    description: echo
                  properties:
                    "0":
                      type: string
                      enum:
                        - pong
                      description: literally
          - name: subscribe
            title: subscribe
            messageId: in//subscribe
            payload:
              type: object
              format: tuple
              additionalProperties:
                format: any
                description: Does not matter
              properties: {}
components: {}
"
`;
